### PR TITLE
Add `is_wrapped` utility function for Fabric

### LINKED ISSUE #16484: Method to check whether a Fabric model was set up already
### Description & Motivation

Currently, if you pass a model to Fabric.setup() that has already been wrapped/set up, you get an error. This is currently expected, because there are several reasons and edge cases where this would be invalid usage. However, the only way to check whether a model is set up already is with an instance check on a protected class:

```py
already_set_up = isinstance(model, _FabricModel)
```

There are valid use case where such a check comes up. For example:

You build a trainer that has a fit method with the following implementation.

```py
model, optimizer = fabric.setup(model, optimizer)
for epoch in range(epochs):
    self.train(model, optimizer)
    self.validate(model)
```

This trainer has a validate() method that can either be called separately, or internally within the fit loop. Hence, the validate() method needs to check whether the model was already set up.

### Pitch

Introduce `Fabric.is_setup()`, or function `is_setup()`.

```py
assert not is_setup(model)
model = fabric.setup(model)
assert is_setup(model)
```

### Alternatives

There are two alternatives:

**Alternative 1**
Make the `Fabric.setup()` a no-op when a model is passed in that was already set up. However, there are problematic edge cases, that break the contract with the strategies. For example, when setup is called with a model that is already set up, but an optimizer that is not set up yet:
```py
model = setup(model)
model = setup(model, optimizer)  # what should happen?
```
OR
If a model is set up twice but with different optimizers
```py
model, optimizer1 = setup(model, optimizer1)
model, optimizer2 = setup(model, optimizer2)
```

**Alternative 2**
The example I gave with the shared validation loop could also be re-written like this:

```py
def train(model, optimizer):
    ...

def validate(model):
    model = fabric.setup(model)
    return self._validate(model)
    
def _validate(model):
    # this assumes the passed in model is already set up

def fit(model):
    model, optimizer = fabric.setup(model, optimizer)
    for epoch in range(epochs):
        self.train(model, optimizer)
        self._validate(model)  # call _validate(), not validate()

```

### Additional context

cc @borda @tchaton @justusschock @awaelchli @carmocca @pietrolesci

### PR DESCRIPTION
**Pull request by Atharva-Phatak**
## What does this PR do?

Method to check whether a Fabric model was set up already #16484 

Fixes #16484 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/fabric.py [hunk 1]
```diff
@@ -799,5 +799,13 @@ def _is_using_cli() -> bool:
     return bool(int(os.environ.get("LT_CLI_USED", "0")))
 
 
+def is_wrapped(module: Union[nn.Module, _FabricModule]) -> bool:
```
Comment (justusschock) [inline src/lightning/fabric/fabric.py:802]: I know this is a very simple function, but could we still get a test for it?

File: src/lightning/fabric/fabric.py [hunk 2]
```diff
@@ -799,5 +799,13 @@ def _is_using_cli() -> bool:
...
+    """Method to check if nn.Module is already wrapped as a FabricModule
+    Args:
+        module: The module to be checked if it is _FabricModule or not
```
Comment (justusschock) [inline src/lightning/fabric/fabric.py:805]: Can we extend this to an arbitrary object and not only check for _FabricModule, but also for _FabricOptimizer etc.?
That would be kind of consistent with `_unwrap_objects`

**Review by justusschock — COMMENTED** [review]
Looks good already. Added a few comments.
Don't forget to add a changelog entry :)

File: src/lightning/fabric/fabric.py (hunk 2 above)
Comment (Atharva-Phatak) [reply src/lightning/fabric/fabric.py:805]: @awaelchli, would you like? function for each check ??

Comment (awaelchli) [reply src/lightning/fabric/fabric.py:805]: No, I think just extending the check inside this function is the best :))

**Review by awaelchli — APPROVED** [review]

File: src/lightning/fabric/wrappers.py [hunk 3]
```diff
@@ -199,3 +199,16 @@ def _unwrap(
         return obj
 
     return apply_to_collection(collection, dtype=(_FabricModule, _FabricOptimizer, _FabricDataLoader), function=_unwrap)
+
+
+def is_wrapped(obj: Union[Any, _FabricModule, _FabricOptimizer, _FabricDataLoader]) -> bool:
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:204]: "Any" makes mypy stop running: https://stackoverflow.com/a/39817126

```suggestion
def is_wrapped(obj: object) -> bool:
```

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
